### PR TITLE
TF-1362: Display report banner

### PR DESCRIPTION
**Pull request by nqhhdev**
TF-1362: Display report banner

- Create Separate Controller for spam report
- Store and Get `lastTimeDismissedSpamReport`

### CODE REVIEW

File: lib/features/mailbox_dashboard/data/model/spam_report_hive_cache.dart [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import 'package:equatable/equatable.dart';
+import 'package:hive/hive.dart';
+import 'package:tmail_ui_user/features/caching/utils/caching_constants.dart';
+
+part 'spam_report_hive_cache.g.dart';
+
+@HiveType(typeId: CachingConstants.SPAM_REPORT_INDENTIRY)
+class SpamReportCache extends HiveObject with EquatableMixin {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/model/spam_report_hive_cache.dart:8]: Should use `share preference` to save `date_time_spam_report`

File: lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/spam_report.dart';
+
+abstract class SpamReportRepository {
+  Future<void> storelastTimeDismissedSpamReported(SpamReport lastTimeDismissedSpamReport);
+
+  Future<SpamReport> getlastTimeDismissedSpamReported();
+}
+
```
Comment (dab246) [inline lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart:8]: remove line

File: lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+
```
Comment (dab246) [inline lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart:10]: same

**Comment by chibenwa** [comment]
If you can record a small video it would be great ;-)

File: lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+class SpamReportCacheManager {
+  final SharedPreferences _sharedPreferences;
+
+  SpamReportCacheManager(this._sharedPreferences);
+
+  Future<int> getlastTimeDismissedSpamReported() async {
+    return _sharedPreferences.getInt(MailboxDashboardConstant.keyLastTimeDismissedSpamReported) ?? 0;
+  }
+
+  Future<void> storelastTimeDismissedSpamReported(int lastTimeDismissedSpamReported) async {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart:13]: should return a `Future`

File: lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+   await _sharedPreferences.setInt(MailboxDashboardConstant.keyLastTimeDismissedSpamReported,lastTimeDismissedSpamReported);
+  }
+
+  Future<void> deleteLastTimeDismissedSpamReported() async {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart:17]: same

File: lib/features/mailbox_dashboard/data/datasource/spam_report_datasource.dart [hunk 6]
```diff
@@ -0,0 +1,7 @@
+abstract class SpamReportDataSource {
+  Future<void> storeLastTimeDismissedSpamReported(int lastTimeDismissedSpamReported);
+
+  Future<int> getLastTimeDismissedSpamReported();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/data/datasource/spam_report_datasource.dart:4]: why not `DateTime`?

File: lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart [hunk 7]
```diff
@@ -0,0 +1,7 @@
+abstract class SpamReportRepository {
+  Future<void> storeLastTimeDismissedSpamReported(int lastTimeDismissedSpamReported);
+
+  Future<int> getLastTimeDismissedSpamReported();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart:4]: Why not `DateTime`?
```suggestion
  Future<DateTime> getLastTimeDismissedSpamReported();
```

**Comment by nqhhdev** [comment]
- Build UI for Web and Responsive


https://user-images.githubusercontent.com/99852347/211368638-334cfd7c-8808-40fb-bd27-0bee4a930097.mov

**Comment by nqhhdev** [comment]
- Build UI for Desktop and Responsive


https://user-images.githubusercontent.com/99852347/211369013-b6f70de7-f880-4f99-9838-495066f10a49.mov

**Comment by nqhhdev** [comment]
- Build UI for Mobile and Responsive


https://user-images.githubusercontent.com/99852347/211369168-87f928a3-3cfa-4dbd-9640-941cdabc07c1.mov

**Comment by hoangdat** [comment]
image

is it ok in your implementation?

File: lib/features/mailbox_dashboard/domain/state/store_spam_report_state.dart [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+
+class StoreSpamReportLoading extends UIState {}
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/state/store_spam_report_state.dart:4]: Store
```suggestion
class StoreLastTimeDismissedSpamReportLoading extends UIState {}
```

Comment (hoangdat) [reply lib/features/mailbox_dashboard/domain/state/store_spam_report_state.dart:4]: idem for other related items :)

File: lib/features/mailbox_dashboard/domain/usecases/get_spam_report_interactor.dart [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:dartz/dartz.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/repository/spam_report_repository.dart';
+
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/state/get_last_time_dismissed_spam_reported_state.dart';
+
+class GetSpamReportInteractor {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/usecases/get_spam_report_interactor.dart:8]:
```suggestion
class GetLastTimeDismissedSpamReportInteractor {
```

File: lib/features/mailbox_dashboard/domain/usecases/store_spam_report_interactor.dart [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:dartz/dartz.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/repository/spam_report_repository.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/state/store_spam_report_state.dart';
+
+class StoreSpamReportInteractor {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/usecases/store_spam_report_interactor.dart:7]:
```suggestion
class StoreLastTimeDismissedSpamReportInteractor {
```

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 11]
```diff
@@ -692,4 +695,76 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
       }
     });
   }
+
+  Widget _buildSpamReportBoxWebWidget(BuildContext context) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:699]: is it ok? I prefer a separated widget for SpamReport. WDYT?

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 12]
```diff
@@ -692,4 +695,76 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
...
+                Row(
+                  children: [
+                    SvgPicture.asset(
+                      imagePaths.icInfoCircleOutline,
+                      width: 28,
+                      height: 28,
+                      color: AppColor.primaryColor,
+                    ),
+                    const SizedBox(
+                      width: 8,
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:728]: If there are not many parameters (<2), they should be written on the same line

**Review by dab246 — COMMENTED** [review]
Please run ` flutter pub run intl_generator` after adding `AppLocalization`

File: test/features/thread/domain/usecases/get_emails_in_mailbox_interactor_test.mocks.dart [hunk 13]
```diff
@@ -109,4 +109,4 @@ class MockThreadRepository extends _i1.Mock implements _i2.ThreadRepository {
               Invocation.method(#emptyTrashFolder, [accountId, trashMailboxId]),
               returnValue: Future<List<_i12.EmailId>>.value(<_i12.EmailId>[]))
           as _i3.Future<List<_i12.EmailId>>);
-}
+}
```
Comment (sherlockvn) [inline test/features/thread/domain/usecases/get_emails_in_mailbox_interactor_test.mocks.dart:112]: remove the changes

File: test/features/thread/domain/usecases/refresh_changes_emails_in_mailbox_interactor_test.mocks.dart [hunk 14]
```diff
@@ -109,4 +109,4 @@ class MockThreadRepository extends _i1.Mock implements _i2.ThreadRepository {
               Invocation.method(#emptyTrashFolder, [accountId, trashMailboxId]),
               returnValue: Future<List<_i12.EmailId>>.value(<_i12.EmailId>[]))
           as _i3.Future<List<_i12.EmailId>>);
-}
+}
```
Comment (sherlockvn) [inline test/features/thread/domain/usecases/refresh_changes_emails_in_mailbox_interactor_test.mocks.dart:112]: remove the changes

**Review by sherlockvn — DISMISSED** [review]

File: lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart [hunk 15]
```diff
@@ -1,7 +1,12 @@
...
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart';
+
+abstract class SpamReportRepository {
   Future<void> storeLastTimeDismissedSpamReported(DateTime lastTimeDismissedSpamReported);
 
   Future<DateTime> getLastTimeDismissedSpamReported();
 
   Future<void> deleteLastTimeDismissedSpamReported();
+
+  Stream<UnreadSpamEmailsResponse> getUnreadSpamEmails(UnreadSpamEmailsRequest unreadSpamEmailsRequest);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart:11]: Can you share the idea why you want to use `Stream` here?

File: lib/features/mailbox_dashboard/domain/model/local_spam_report.dart [hunk 16]
```diff
@@ -1,10 +1,10 @@
 import 'package:equatable/equatable.dart';
 
-class SpamReport with EquatableMixin {
+class LocalSpamReport with EquatableMixin {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/model/local_spam_report.dart:3]: what use for?

File: lib/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart [hunk 17]
```diff
@@ -0,0 +1,14 @@
+
+import 'package:equatable/equatable.dart';
+import 'package:jmap_dart_client/jmap/mail/email/email.dart';
+
+class UnreadSpamEmailsResponse with EquatableMixin {
+  final List<Email>? spamEmailList;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart:6]: IMO, we will not receive `spamEmailList`. Moreover, we only care for the mailbox with `spam` role

File: lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart [hunk 18]
```diff
@@ -0,0 +1,23 @@
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/utils/mailbox_dashboard_constant.dart';
+
+class SpamReportCacheManager {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/data/local/spam_report_cache_manager.dart:4]: Should be an 
```suggestion
class SharePreferenceSpamReportDataSource extend xxxDataSource {
```

File: lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart [hunk 19]
```diff
@@ -0,0 +1,31 @@
+import 'package:core/data/model/source_type/data_source_type.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/data/datasource/spam_report_datasource.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_request.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/repository/spam_report_repository.dart';
+
+class SpamReportRepositoryImpl extends SpamReportRepository {
+  final Map<DataSourceType, SpamReportDataSource> mapDataSource;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart:8]: final `SpamReportDataSource`

File: lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+   return mapDataSource[DataSourceType.local]!.storeLastTimeDismissedSpamReported(lastTimeDismissedSpamReported);
+  }
+  
+  @override
+  Future<void> deleteLastTimeDismissedSpamReported() {
+    return mapDataSource[DataSourceType.local]!.deleteLastTimeDismissedSpamReported();
+  }
+
+  @override
+  Stream<UnreadSpamEmailsResponse> getUnreadSpamEmails(UnreadSpamEmailsRequest unreadSpamEmailsRequest) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart:28]: How about `Future`?

File: lib/features/mailbox_dashboard/domain/model/unread_spam_emails_request.dart [hunk 21]
```diff
@@ -0,0 +1,16 @@
+
+import 'package:equatable/equatable.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/unsigned_int.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox_filter_condition.dart';
+
+class UnreadSpamEmailsRequest with EquatableMixin {
+  final AccountId accountId;
+  final UnsignedInt? limit;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/model/unread_spam_emails_request.dart:9]: should use with default value = 1

File: lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart';
+
+abstract class SpamReportRepository {
+  Future<void> storeLastTimeDismissedSpamReported(DateTime lastTimeDismissedSpamReported);
+
+  Future<DateTime> getLastTimeDismissedSpamReported();
+
+  Future<void> deleteLastTimeDismissedSpamReported();
+
+  Stream<UnreadSpamEmailsResponse> getUnreadSpamEmails(UnreadSpamEmailsRequest unreadSpamEmailsRequest);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/repository/spam_report_repository.dart:11]: `Future`

File: lib/features/mailbox_dashboard/domain/state/get_unread_spam_emails_state.dart [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+class GetUnreadSpamEmailsLoading extends UIState {}
+
+class GetUnreadSpamEmailsSuccess extends UIState {
+  final int unreadSpamEmailNumber;
+
+  GetUnreadSpamEmailsSuccess(this.unreadSpamEmailNumber);
+
+  @override
+  List<Object> get props => [];
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/state/get_unread_spam_emails_state.dart:12]:
```suggestion
  List<Object> get props => [unreadSpamEmailNumber];
```

File: lib/features/mailbox_dashboard/domain/usecases/get_unread_spam_emails_interactor.dart [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+import 'package:dartz/dartz.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/unsigned_int.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox_filter_condition.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_request.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/repository/spam_report_repository.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/state/get_unread_spam_emails_state.dart';
+
+class GetUnreadSpamEmailsInteractor {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/usecases/get_unread_spam_emails_interactor.dart:13]: Change the name of use-case, UI state, failure

`InvalidSpamReportCondition extend FeatureFailure`

**Comment by nqhhdev** [comment]
- Demo for Web


https://user-images.githubusercontent.com/99852347/212244279-488fad8e-0625-4245-95eb-7d9d2f17ca47.mov

**Comment by nqhhdev** [comment]
Demo for Desktop

https://user-images.githubusercontent.com/99852347/212244732-68a02aac-334d-46f7-9892-8b85fbc7957d.mov

**Comment by nqhhdev** [comment]
Demo for Mobile 


https://user-images.githubusercontent.com/99852347/212245077-0b1259fd-b3b5-430c-9515-a13b2a42e38e.mov

File: lib/features/mailbox_dashboard/data/datasource_impl/share_preference_spam_report_data_source_impl.dart [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:tmail_ui_user/features/mailbox_dashboard/data/datasource/spam_report_datasource.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/data/local/share_preference_spam_report_data_source.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/model/unread_spam_emails_response.dart';
+import 'package:tmail_ui_user/main/exceptions/exception_thrower.dart';
+
+class SharePreferenceSpamReportDataSourceImpl extends SpamReportDataSource {
+  final SharePreferenceSpamReportDataSource _sharePreferenceSpamReportDataSource;
+  final ExceptionThrower _exceptionThrower;
+
+  SharePreferenceSpamReportDataSourceImpl(this._sharePreferenceSpamReportDataSource, this._exceptionThrower);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/datasource_impl/share_preference_spam_report_data_source_impl.dart:13]: remove line

File: lib/features/mailbox_dashboard/data/network/spam_report_api.dart [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+          .execute();
+
+    final _mailboxResponse = result
+        .parse<GetMailboxResponse>(getMailboxInvocation.methodCallId, GetMailboxResponse.deserialize);
+    
+    final _unreadSpamMailbox = _mailboxResponse?.list.first;
+
+     return Future.sync(() async {
+      return UnreadSpamEmailsResponse(unreadSpamMailbox: _unreadSpamMailbox);
+    }).catchError((error) {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/network/spam_report_api.dart:51]: Using like that will not catch the error. It should be written like that.
```suggestion
      return Future.sync(() async {
         final _unreadSpamMailbox = _mailboxResponse?.list.first;
       return UnreadSpamEmailsResponse(unreadSpamMailbox: _unreadSpamMailbox);
     }).catchError((error) {
       throw error;
     });
```

File: lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+  @override
+  Future<UnreadSpamEmailsResponse> getUnreadSpamMailbox(
+    AccountId accountId,
+    {
+      MailboxFilterCondition? mailboxFilterCondition,
+      UnsignedInt? limit
+    }
+  ) {
+    return mapDataSource[DataSourceType.network]!.findNumberOfUnreadSpamEmails(
+      accountId, mailboxFilterCondition: mailboxFilterCondition, limit: limit);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart:38]: each parameters should at one line

File: lib/features/mailbox_dashboard/domain/usecases/get_unread_spam_mailbox_interactor.dart [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+  ) async* {
+    try {
+      yield Right(GetUnreadSpamMailboxLoading());
+      final _lastTimeDissmissedSpamReported = await _spamReportRepository.getLastTimeDismissedSpamReported();
+      final _timeLast = DateTime.now().difference(_lastTimeDissmissedSpamReported);
+
+      final _checkTimeCondition = (_timeLast.inHours > 0) && (_timeLast.inHours > conditionsForDisplayingSpamReportBanner);
+
+      if (_checkTimeCondition) {
+        final _response =  await _spamReportRepository.getUnreadSpamMailbox(accountId, mailboxFilterCondition: mailboxFilterCondition, limit: limit);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/domain/usecases/get_unread_spam_mailbox_interactor.dart:32]: same

File: lib/features/mailbox_dashboard/presentation/bindings/mailbox_dashboard_bindings.dart [hunk 29]
```diff
@@ -185,6 +201,14 @@ class MailboxDashBoardBindings extends BaseBindings {
       Get.find<RemoteExceptionThrower>(),
     ));
     Get.lazyPut(() => SessionStorageComposerDatasourceImpl());
+     Get.lazyPut(() => SpamReportDataSourceImpl(
+      Get.find<SpamReportApi>(),
+      Get.find<RemoteExceptionThrower>(),
+    ));    
+    Get.lazyPut(() => SharePreferenceSpamReportDataSourceImpl(
+      Get.find<SharePreferenceSpamReportDataSource>(),
+      Get.find<RemoteExceptionThrower>(),
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/bindings/mailbox_dashboard_bindings.dart:210]: This datasource handles cache-side data. Should use `CacheExceptionThower`

File: lib/features/thread/presentation/thread_controller.dart [hunk 30]
```diff
@@ -119,6 +120,8 @@ class ThreadController extends BaseController with EmailActionController {
 
   RxList<PresentationEmail> get emailList => mailboxDashBoardController.emailsInCurrentMailbox;
 
+  SpamReportController get spamReportController => mailboxDashBoardController.spamReportController;
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:123]: why thread controller need to ref to spamReportController?

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 31]
```diff
@@ -1400,6 +1403,8 @@ class MailboxDashBoardController extends ReloadableController {
   void popAllRouteIfHave() {
     Get.until((route) => Get.currentRoute == AppRoutes.dashboard);
   }
+
+  void getNumberOfUnreadSpamMailsAction() => spamReportController.getUnreadSpamMailboxAction(accountId.value!);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1407]: Where it used?

File: lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+  
+  void dismissSpamReportAction() {
+    _storeLastTimeDismissedSpamReportedAction();
+  }
+
+  void getUnreadSpamMailboxAction(AccountId accountId){
+    try {
+      final _mailboxFilterCondition = MailboxFilterCondition(role: Role('Spam'));
+      consumeState(_getNumberOfUnreadSpamEmailsInteractor.execute(accountId,mailboxFilterCondition: _mailboxFilterCondition));
+    } catch(e) {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart:50]: IMO, It is not necessary to use `try/catch` here. 
Note: Only use `try/catch` in a function where we are not sure if the values of the variables used are null or not?

File: lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+      logError('SpamReportController::getUnreadSpamMailboxAction(): exception: $e');
+    }
+  }
+
+  void _storeLastTimeDismissedSpamReportedAction() {
+    try {
+      consumeState(_storeSpamReportInteractor.execute(DateTime.now()));
+    } catch(e) {
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart:58]: same

File: lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart [hunk 34]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  bool get notShowSpamReportBanner => _presentationSpamMailbox.value == null;
+
+  int get getNumberOfUnreadSpamEmails => (_presentationSpamMailbox.value?.unreadEmails?.value.value ?? 0).toInt();
+
+  void openMailbox(
+      BuildContext context,
+  ) {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart:69]:
```suggestion
  void openMailbox(BuildContext context) {
```

File: lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+
+  final _presentationSpamMailbox = Rxn<PresentationMailbox>();
+
+  SpamReportController(this._storeSpamReportInteractor, this._getNumberOfUnreadSpamEmailsInteractor);
+
+  @override
+  void onDone() {
+     viewState.value.fold(
+      (failure) {
+        logError('SpamReportController::SpamReportControllerFailure(): exception: $failure');
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart:28]:
```suggestion
        logError('SpamReportController::onDone(): failure: $failure');
```

File: lib/features/mailbox_dashboard/presentation/action/dashboard_action.dart [hunk 36]
```diff
@@ -123,4 +124,14 @@ class SearchEmailByQueryAction extends DashBoardAction {
 
   @override
   List<Object?> get props => [navigationRouter];
+}
+class OpenSpamMailboxAction extends DashBoardAction {
+
+  final BuildContext context;
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/action/dashboard_action.dart:130]: what? first time to see it in action :))))

File: lib/features/thread/presentation/thread_controller.dart [hunk 37]
```diff
@@ -117,6 +118,8 @@ class ThreadController extends BaseController with EmailActionController {
 
   SearchQuery? get searchQuery => searchController.searchEmailFilter.value.text;
 
+  RxList<PresentationEmail> get emailList => mailboxDashBoardController.emailsInCurrentMailbox;
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:121]: IMO, we can remove it

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 38]
```diff
@@ -1518,6 +1521,10 @@ class MailboxDashBoardController extends ReloadableController {
   void updateEmailList(List<PresentationEmail> newEmailList) {
     emailsInCurrentMailbox.value = newEmailList;
   }
+  
+  void openSpamMailboxAction(BuildContext context, PresentationMailbox presentationMailbox){
+    dashBoardAction.value = OpenSpamMailboxAction(context, presentationMailbox);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1526]: You can use `dispatchAction` method to update `dashBoardAction`

```suggestion
dispatchAction(OpenSpamMailboxAction(context, presentationMailbox));
```

File: lib/features/mailbox_dashboard/presentation/widgets/spam_report_banner_web_widget.dart [hunk 39]
```diff
@@ -0,0 +1,87 @@
...
+import 'package:core/presentation/resources/image_paths.dart';
+import 'package:core/presentation/views/button/icon_button_web.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_svg/flutter_svg.dart';
+import 'package:get/get.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart';
+import 'package:tmail_ui_user/main/localizations/app_localizations.dart';
+
+class SpamReportBannerWebWidget extends StatelessWidget {
+const SpamReportBannerWebWidget({ Key? key }) : super(key: key);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/widgets/spam_report_banner_web_widget.dart:11]: add indent

File: lib/features/thread/presentation/thread_controller.dart [hunk 40]
```diff
@@ -34,6 +34,7 @@ import 'package:tmail_ui_user/features/mailbox/domain/state/mark_as_mailbox_read
 import 'package:tmail_ui_user/features/mailbox_dashboard/domain/state/remove_email_drafts_state.dart';
 import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/action/dashboard_action.dart';
 import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/controller/search_controller.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart';
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:37]: hmm

File: lib/features/thread/presentation/thread_controller.dart [hunk 41]
```diff
@@ -117,6 +118,7 @@ class ThreadController extends BaseController with EmailActionController {
 
   SearchQuery? get searchQuery => searchController.searchEmailFilter.value.text;
 
+
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:121]: no need

**Review by dab246 — APPROVED** [review]

File: lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart [hunk 42]
```diff
@@ -27,14 +27,13 @@ class SpamReportRepositoryImpl extends SpamReportRepository {
...
+  Future<UnreadSpamEmailsResponse> getUnreadSpamMailbox(
     AccountId accountId,
     {
       MailboxFilterCondition? mailboxFilterCondition,
       UnsignedInt? limit
     }
   ) {
-    return mapDataSource[DataSourceType.network]!.findNumberOfUnreadSpamEmails(
-      accountId, mailboxFilterCondition: mailboxFilterCondition, limit: limit);
+    return mapDataSource[DataSourceType.network]!.findNumberOfUnreadSpamEmails(accountId, mailboxFilterCondition: mailboxFilterCondition, limit: limit);
```
Comment (sherlockvn) [inline lib/features/mailbox_dashboard/data/repository/spam_report_repository_impl.dart:37]: should make each argument in each line

File: lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart [hunk 43]
```diff
@@ -1,14 +1,64 @@
...
+    consumeState(_getNumberOfUnreadSpamEmailsInteractor.execute(accountId,mailboxFilterCondition: _mailboxFilterCondition));
+  }
+
+  void _storeLastTimeDismissedSpamReportedAction() {
+    consumeState(_storeSpamReportInteractor.execute(DateTime.now()));
+  }
+
+  bool get notShowSpamReportBanner => _presentationSpamMailbox.value == null;
+
+  int get getNumberOfUnreadSpamEmails => (_presentationSpamMailbox.value?.unreadEmails?.value.value ?? 0).toInt();
```
Comment (sherlockvn) [inline lib/features/mailbox_dashboard/presentation/controller/spam_report_controller.dart:57]: IMO, `getNumberOfUnreadSpamEmails` should be `numberOfUnreadSpam` or `numberOfUnreadSpamEmails`

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 44]
```diff
@@ -107,7 +108,7 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
                     children: [
                       SizedBox(child: MailboxView(), width: responsiveUtils.defaultSizeMenu),
                       Expanded(child: Column(children: [
-                        _buildSpamReportBoxWebWidget(context),
+                        const SpamReportBannerWebWidget(),
                         _buildEmptyTrashButton(context),
                         const QuotasWarningBannerWidget(
                           margin: EdgeInsets.only(right: 16, top: 8),
```
Comment (sherlockvn) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:114]: make it `const EdgeInsets.only(right: 16, top: 8)`

**Review by sherlockvn — APPROVED** [review]

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 44 above)
Comment (nqhhdev) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:114]: const QuotasWarningBannerWidget(
                          margin: EdgeInsets.only(right: 16, top: 8),  
                          
:)))

**Review by hoangdat — APPROVED** [review]

**Pull request merged**